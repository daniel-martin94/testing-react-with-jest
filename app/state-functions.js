import shortid from 'shortid';

export function toggleDone(state, id) {
  const todos = state.todos.map((todo) => {
    if (todo.id === id) {
      todo.done = !todo.done;
    }

    return todo;
  });

  return { todos };
}

export function addTodo(state, todo) {
  const newTodo = Object.assign({}, todo, {
    id: shortid.generate(),
    done: false
  });
  
  return {
    todos: state.todos.concat([newTodo])
  };
}

export function deleteTodo(state, id) {
 
  let newDeleted = state.todos.filter(todo => {
    if (todo.id === id) {
      return todo
    }
  })
  return {
    todos: state.todos.filter((todo) => todo.id !== id),
    deletedTodos: state.deletedTodos.concat(newDeleted)
  };
}

export function undoDelete(state) {
  let lastDeleted = state.deletedTodos.pop()
  return {
    todos: state.todos.concat([lastDeleted]),
    deletedTodos: state.deletedTodos
  };
}
