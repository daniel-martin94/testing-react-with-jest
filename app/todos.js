import React, { useState } from 'react';
import Todo from './todo';
import AddTodo from './add-todo';
import UndoTodo from './undo-delete';
import shortid from 'shortid';

import {
  toggleDone,
  addTodo,
  deleteTodo,
  undoDelete
} from './state-functions';

export default class Todos extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      todos: [
        { id: shortid.generate(), name: 'Write a blog post for Sitepoint', done: false },
        { id: shortid.generate(), name: 'Blog about Jest', done: false },
        { id: shortid.generate(), name: 'Walk the dog', done: false },
      ],
      deletedTodos: []
    }
  }
  toggleDone(id) {
    this.setState(toggleDone(this.state, id));
  }

  addTodo(todo) {
    this.setState(addTodo(this.state, todo));
  }

  deleteTodo(id) {
    this.setState(deleteTodo(this.state, id));
  }

  undoDelete() {
    this.setState(undoDelete(this.state))
  }

  renderTodos() {
    return this.state.todos.map((todo) => {
      return (
        <li key={todo.id}>
          <Todo
            todo={todo}
            doneChange={(id) => this.toggleDone(id)}
            deleteTodo={(id) => this.deleteTodo(id)} />
        </li>
      );
    });
  }

  render() {
    let { deletedTodos }  = this.state
    return (
      <div>
        <div>
          <p>The <em>best</em> todo app out there.</p>
          { deletedTodos.length > 0 && 
            <UndoTodo undo={() => this.undoDelete()}></UndoTodo>
          }
        </div>
        <h1>Things to get done:</h1>
        <ul className="todos-list">{ this.renderTodos() }</ul>
        <AddTodo onNewTodo={(todo) => this.addTodo(todo)} />
      </div>
    )
  }
}
