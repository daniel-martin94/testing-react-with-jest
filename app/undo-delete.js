import React, { useState, Component } from 'react'
import { render } from 'react-dom'

const undoDelete = (props) => {
    let { undo } = props
    return (
        <div>
            <input type="button" value="Undo Delete" onClick={undo}/>
        </div>
    )
}

export default undoDelete