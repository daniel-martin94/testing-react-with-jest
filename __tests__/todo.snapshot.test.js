import Todo from '../app/todo';
import UndoTodo from '../app/undo-delete'
import renderer from 'react-test-renderer';
import React from 'react';
import { undoDelete } from '../app/state-functions'
jest.mock('react-dom')

test('Todo component renders correctly', () => {
    const todo = { id: 1, done: false, name: 'Buy Milk' };
    const rendered = renderer.create(
      <Todo todo={todo}> </Todo>
    )
    expect(rendered.toJSON()).toMatchSnapshot()
  })

test('Test for undo delete button', () => {
  const startState = {
    todos: [],
    deletedTodos: [{ id: 1, done: false, text: 'Buy Milk' }]
  };
  const rendered = renderer.create(
      <UndoTodo undo={undoDelete}> </UndoTodo>
  )
  expect(rendered.toJSON()).toMatchSnapshot()
})
