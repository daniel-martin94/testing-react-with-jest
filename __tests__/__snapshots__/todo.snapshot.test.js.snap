// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for undo delete button 1`] = `
<div>
  <input
    onClick={[Function]}
    type="button"
    value="Undo Delete"
  />
</div>
`;

exports[`Todo component renders correctly 1`] = `
<div
  className="todo  todo-1"
>
  <p
    className="toggle-todo"
    onClick={[Function]}
  >
    Buy Milk
  </p>
  <a
    className="delete-todo"
    href="#"
    onClick={[Function]}
  >
    Delete
  </a>
</div>
`;
